WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift

```diff
@@ -475,85 +475,80 @@ class AbstractPostListViewController: UIViewController,
+        let result = try await repository.paginate(
+            type: postType == .post ? Post.self : Page.self,
+            statuses: filter.statuses,
+            authorUserID: author,
+            offset: isFirstPage ? 0 : fetchResultsController.fetchedObjects?.count ?? 0,
+            number: number,
+            in: blogID
+        )
+
+        let posts = try result.map { try coreDataStack.mainContext.existingObject(with: $0) }
+        let hasMore = result.count >= number
+
+        return (posts, hasMore)
+    }
+
+    func syncHelper(_ syncHelper: WPContentSyncHelper, syncContentWithUserInteraction userInteraction: Bool, success: ((_ hasMore: Bool) -> ())?, failure: ((_ error: NSError) -> ())?) {
+        let filter = filterSettings.currentPostListFilter()
+        Task { @MainActor [weak self] in
+            do {
+                guard let (posts, hasMore) = try await self?.syncPosts(isFirstPage: true) else { return }
+
+                guard let self else { return }
 
                 if posts.count > 0 {
-                    self.updateFilter(filter, withSyncedPosts: posts, syncOptions: options)
+                    self.updateFilter(filter, withSyncedPosts: posts, hasMore: hasMore)
                     SearchManager.shared.indexItems(posts)
                 }
 
                 success?(filter.hasMore)
-            }, failure: { [weak self] error in
-
-                guard let self, let error else {
-                    return
-                }
+            } catch {
+                guard let self else { return }
 
                 failure?(error as NSError)
 
                 if userInteraction == true {
                     self.handleSyncFailure(error as NSError)
                 }
-        })
+            }
+        }
     }
 
     func syncHelper(_ syncHelper: WPContentSyncHelper, syncMoreWithSuccess success: ((_ hasMore: Bool) -> Void)?, failure: ((_ error: NSError) -> Void)?) {
 
         setFooterHidden(false)
 
-        let postType = postTypeToSync()
         let filter = filterSettings.currentPostListFilter()
-        let author = filterSettings.shouldShowOnlyMyPosts() ? blogUserID() : nil
 
-        let postService = PostService(managedObjectContext: managedObjectContext())
-
-        let options = PostServiceSyncOptions()
-        options.statuses = filter.statuses.strings
-        options.authorID = author
-        options.number = numberOfLoadedElement
-        options.offset = fetchResultsController.fetchedObjects?.count as NSNumber?
-
-        postService.syncPosts(
-            ofType: postType,
-            with: options,
-            for: blog,
-            success: { [weak self] posts in
-                guard let self, let posts else {
-                    return
-                }
+        Task { @MainActor [weak self] in
+            do {
+                guard let (posts, hasMore) = try await self?.syncPosts(isFirstPage: false) else { return }
+
+                guard let self else { return }
```

Comment:
Would it make sense to swap these? The end result would be the same, return if self is not available, but I feel it's odd to access `self?` and check `let self` afterwards. 🤷‍♂️ 

```suggestion
                guard let self else { return }

                let (posts, hasMore) = try await self.syncPosts(isFirstPage: false)
```

I didn't try this code, maybe I'm missing something and in fact it's necessary to do `guard` twice.

Reply:
This `guard let self` is similar to how we do the same check in a completion block. In reality, `self` should never be nil when calling `syncPosts`, but could be nil if user has exit the screen before the `sycnPosts` returns. I have added a comment in https://github.com/wordpress-mobile/WordPress-iOS/pull/22070/commits/19307ccad85c448547c6a84987894cdaecfd78fb